src/os/ObjectStore.h

```diff
@@ -1718,22 +1718,127 @@ class ObjectStore {
-    const coll_t& cid,
+  virtual int read(
+   const coll_t& cid,
+   const ghobject_t& oid,
+   uint64_t offset,
+   size_t len,
+   bufferlist& bl,
+   uint32_t op_flags = 0) = 0;
+  virtual int read(
+    CollectionHandle &c,
     const ghobject_t& oid,
     uint64_t offset,
     size_t len,
     bufferlist& bl,
-    uint32_t op_flags = 0) = 0;
-   virtual int read(
-     CollectionHandle &c,
-     const ghobject_t& oid,
-     uint64_t offset,
-     size_t len,
-     bufferlist& bl,
-     uint32_t op_flags = 0) {
-     return read(c->get_cid(), oid, offset, len, bl, op_flags);
-   }
+    uint32_t op_flags = 0) {
+    return read(c->get_cid(), oid, offset, len, bl, op_flags);
+  }
+
+  /*********************************
+   * ReadTransaction
+   *
+   * A Transaction represents a sequence of primitive non-mutating
+   * operations over the same ghobject in the same collection.
+   */
+  class ReadTransaction {
+  public:
+    virtual ~ReadTransaction() = default;
+
+    /**
+     * read -- read a byte range of data from an object.
+     *
+     * The operation can be performed synchronously or asynchronously
+     * solely depnding on the implementation's decision.
+     *
+     * Note: if reading from an offset past the end of the object, we
+     * return 0 (not, say, -EINVAL).
+     *
+     * @param offset location offset of first byte to be read
+     * @param len number of bytes to be read
+     * @param op_flags is CEPH_OSD_OP_FLAG_*
+     * @param destbl output bufferlist
+     * @param on_complete callback to be called
+     * @returns number of bytes read on success,
+     *          or negative error code on failure,
+     *          or -EINPROGESS if implementations decided to go async.
+     */
+    virtual int read(
+      uint64_t offset,
+      uint64_t length,
+      uint32_t flags,
+      ceph::bufferlist& destbl,
+      Context* on_complete) = 0;
+
+    /**
+     * read_sparse -- read a byte range of data from an object using
+     * the formatting of CEPH_OSD_OP_SPARSE_READ for output.
+     *
+     * The operation can be performed synchronously or asynchronously
+     * solely depnding on the implementation's decision.
+     *
+     * Note: if reading from an offset past the end of the object, we
+     * return 0 (not, say, -EINVAL).
+     *
+     * @param offset location offset of first byte to be read
+     * @param len number of bytes to be read
+     * @param op_flags is CEPH_OSD_OP_FLAG_*
+     * @param destbl output bufferlist
+     * @param on_complete callback to be called
+     * @returns number of bytes read on success,
+     *          or negative error code on failure,
+     *          or -EINPROGESS if implementations decided to go async.
+     */
+    virtual int read_sparse(
+      uint64_t offset,
+      uint64_t length,
+      uint32_t flags,
+      ceph::bufferlist& destbl,
+      Context* on_complete) = 0;
+
+    virtual bool empty() const = 0;
+    virtual int apply(Context* on_complete) = 0;
```

Comment:
Hmm, having the `::apply` as a part of `ObjectStore` and taking `ReadTransaction` as `std::unique_ptr<ReadTransaction>&&` would allow to save separate allocation of `BlueStore::AioReadBatch`.